aptos-move/framework/supra-framework/sources/multisig_voting.spec.move

```diff
@@ -0,0 +1,427 @@
+        aborts_if !std::string::spec_internal_check_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        let execution_key = std::string::spec_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        aborts_if simple_map::spec_contains_key(proposal.metadata, execution_key) &&
+            simple_map::spec_get(proposal.metadata, execution_key) != std::bcs::serialize(false);
+        // aborts_if if (should_pass) { proposal.yes_votes + num_votes > MAX_U128 } else { proposal.no_votes + num_votes > MAX_U128 };
+
+        aborts_if !std::string::spec_internal_check_utf8(RESOLVABLE_TIME_METADATA_KEY);
+
+        let post post_voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let post post_proposal = table::spec_get(post_voting_forum.proposals, proposal_id);
+        // ensures if (should_pass) {
+        //     post_proposal.yes_votes == proposal.yes_votes + num_votes
+        // } else {
+        //     post_proposal.no_votes == proposal.no_votes + num_votes
+        // };
+        let timestamp_secs_bytes = std::bcs::serialize(timestamp::spec_now_seconds());
+        let key = std::string::spec_utf8(RESOLVABLE_TIME_METADATA_KEY);
+        ensures simple_map::spec_get(post_proposal.metadata, key) == timestamp_secs_bytes;
+    }
+
+    spec is_proposal_resolvable {
+        use supra_framework::chain_status;
+        // Ensures existence of Timestamp
+        requires chain_status::is_operating();
+        //TODO: Remove pragma aborts_if_is_partial;
+        pragma aborts_if_is_partial = true;
+        // include IsProposalResolvableAbortsIf<ProposalType>;
+    }
+
+    spec resolve {
+        use supra_framework::chain_status;
+        // Ensures existence of Timestamp
+        requires chain_status::is_operating();
+        //TODO: Remove pragma aborts_if_is_partial;
+        pragma aborts_if_is_partial = true;
+        // include IsProposalResolvableAbortsIf<ProposalType>;
+        aborts_if !std::string::spec_internal_check_utf8(IS_MULTI_STEP_PROPOSAL_KEY);
+        let voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::spec_get(voting_forum.proposals, proposal_id);
+        let multi_step_key = std::string::spec_utf8(IS_MULTI_STEP_PROPOSAL_KEY);
+        let has_multi_step_key = simple_map::spec_contains_key(proposal.metadata, multi_step_key);
+        aborts_if has_multi_step_key && !from_bcs::deserializable<bool>(simple_map::spec_get(proposal.metadata, multi_step_key));
+        aborts_if has_multi_step_key && from_bcs::deserialize<bool>(simple_map::spec_get(proposal.metadata, multi_step_key));
+
+        let post post_voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let post post_proposal = table::spec_get(post_voting_forum.proposals, proposal_id);
+        aborts_if !exists<timestamp::CurrentTimeMicroseconds>(@supra_framework);
```

Comment:
What is the role of this?

Reply:
This is for checking the whether time already started. This is required if we have call to timestamp::now_seconds()